File: 0410-split-array-largest-sum/0410-split-array-largest-sum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var splitArray = function(nums, k) {
    function ispossible(mid, nums, k){
        let sum=0;
        let c=1;
        for(let i=0;i<nums.length;i++){
            if (nums[i]>mid){
                return false
            }
            if(sum+nums[i]<=mid){
                sum = sum + nums[i];

            }
            else{
                c++;
                sum = nums[i];
                if(c>k){
                    return false
                }
            }

        }
        return true
    }


    let low = Math.max(...nums);
    let high = nums.reduce((acc,current)=>acc+current,0);
    let res=-1;
    while(low<=high){
        let mid = Math.floor((low+high)/2);
        if (ispossible(mid, nums, k)){
            res = mid;
            high = mid-1;
        }
        else{
            low = mid+1;
        }
    }
    return res;
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    bool calculatesubarrays(vector<int>& nums, int mid, int k) {
        int calculatesubarrays = 1;
        int currentsum = 0;
        for (int i = 0; i < nums.size(); i++) {
            if(nums[i]>mid){
                return false;
            }
           
            if (currentsum + nums[i] <= mid) {
                currentsum += nums[i];
            } else {
                currentsum = nums[i];
                calculatesubarrays++;
            }
        }
        
        if (calculatesubarrays <= k) return true;

        return false;
    }
    int splitArray(vector<int>& nums, int k) {

        int low = INT_MAX;
        int high = 0;
        int ans = -1;

        for (int i = 0; i < nums.size(); i++) {
            high += nums[i];
            low = min(nums[i], low);
        }
        while (low <= high) {
            int mid = (low + high) / 2;
            // cout<<mid<<endl;

            if (calculatesubarrays(nums, mid, k)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }
};